### ISSUE TITLE
openDBChannel & patch not producing consistent results when updating an object field

### ISSUE THREAD

**Issue by yungrich** [mesqueeb/vuex-easy-firestore#180]
Hi I've recently started using your plugin on a new project and I must say I've was loving it when I first started. But I've recently come across a use case where I've spent quite a few hours trying to understand. 

I have a user module setup in 'doc mode, and I do the standard `store.dispatch('userData/openDBChannel')` inside of the `onAuthStateChanged` and that loads all the user data perfectly fine

Then in my component I have 2 functions, to Join and Leave a "game"
```javascript
join () {
  const games = {}
  games[this.$route.params.id] = {
    title: 'Game Title'
  }
  this.$store.dispatch('userData/patch', {
    games,
    gameCounter: 1
  })
},
leave () {
  this.$store.dispatch('userData/patch', {
    games: {},
    gameCounter: 0
  })
}
```
Now this is where things get weird. The `gameCounter` field updates correctly in the vuex store, as does both the `games` and `gameCounter` fields in the firestore db. But the games field does a few weird things in the vuex store. 
I'll refer to the computer I'm coding on as the `local device` and, I'm testing the real time updates on another device as well which I'll call the `remote device`
1) the `games` object does not change at all in the vuex state on the `local device` when switching between Join and Leave
2) the `games` object updates on the `remote device`, but it leaves the key value there with an empty object for example
```javascript
games: {
  yFKbLktpoQYDcflfXt3VPgeFOum1: {}
}
```
whereas it should just be an empty object as it is in the firestore db

Then when I write my own actions and replace `openDBChannel` with firestore's `.onSnapshot()`, and I replace `patch` with firestore's `.update()`, I get the results I was expecting. Everything perfectly syncs with the firestore db on both local and remote devices

I'm on verge of refactoring my codebase to remove all instances of your plugin to avoid such headaches, but was hoping you could shed some light on something I may be doing wrong.

**Comment by mesqueeb** [comment]
@yungrich I think I understand what you mean. This is the expected behaviour because of how changes are merged with the existing results.

> The default behaviour is when a document comes into your app via `openDBChannel` from Firebase: **objects are merged, other values are overwritten.**

### Example of current behaviour

I will give an example. Say you have this document in your vuex:
```js
games: {
  cards: {title: 'cards', peeps: 1, entered: new Date()}
}
```
Your vuex-easy-firestore settings have:
```js
sync: {
  guard: ['games.*.entered']
}
```
This makes sure that the field `entered` is a local value only which is available in your vuex module but will not be synced to Firestore.

With this setup, when a change on another client changes something and Firestore `onSnapshot` listener pushes this object:
```js
games: {
  cards: {title: 'cards', peeps: 2}
}
```
As you can see `peeps` is incremented.

‼️ **At this point we don't want to overwrite the entire `cards` object, because we would loose all guarded values in my vuex store** (like `entered`).
→　That's why my `openDBChannel` function will merge objects of incoming docs.

However, this also means that even when another device sets `games` to `{}` then all other devices will keep their contents, because objects are merged.

## Solution

It's really easy to solve this, just by setting `games` to `null` instead of an empty object. I believe you would only need to make small changes to your UI to make sure functions that expect an object first check if it's `null` or not.

You could also make a getter in your same vuex module like so:
```js
games (state) {
  return (state.games === null) ? {} : state.games
}
```

 If this solved your issue please close it. 
And feel free to continue the discussion below! 😊

**Comment by yungrich** [comment]
@mesqueeb Thank you for the quick reply! It's exactly what's happening. #73 was the same issue wish I had seen it sorry about that. Almost there...

Using the delete action is working much better
```javascript
const gameId = yFKbLktpoQYDcflfXt3VPgeFOum1
this.$store.dispatch('userData/delete', `games.${gameId}`)
```
Everything is working as expected on the `local device`. However I'm still facing the same issue on the `remote device` after it syncs, i get:
```javascript
games: {
  yFKbLktpoQYDcflfXt3VPgeFOum1: {},
  OZiFuNr0i5E9fYL6LYtR: { title: 'game title' }
}
```
instead of: (which is on the `local device`)
```javascript
games: {
  OZiFuNr0i5E9fYL6LYtR: { title: 'game title' }
}
```
Thanks for your help!

**Comment by mesqueeb** [comment]
Hi @yungrich 
i’ll look into a way to prevent that. Please hold on a few hours!!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by mesqueeb** [comment]
@yungrich It's weird because I had already fixed this error, so I'm not sure why it's not working on your side.

Are you using 'doc' or 'collection' mode?

**Comment by mesqueeb** [comment]
@yungrich is the value that you are trying to delete also a [default value](https://mesqueeb.github.io/vuex-easy-firestore/extra-features.html#default-values)?

**Comment by louisameline** [comment]
I think this issue actually needs to be properly solved.
When the data comes from openDBChannel, the library should take it, add the local "guarded properties" to it, and then replace the old data.
The data coming from Firestore is the one source of truth, the local store should be a strict representation of it.

**Comment by louisameline** [comment]
Ok, I looked at the code. Unlike what is implied in the second message of this thread, there is a function that tries to delete local values that are not in the remote document.

However it fails becauses it compares flattened values.

Local:
```
{ a: {b: {c: 1 }}}
// flattens to `a.b.c`
```
Remote:
```
{ a: {}}
// flattens to
`a`
```
So the library thinks "`a.b.c` is not present anymore" and it does `delete a.b.c`, so we're left with `{a: b: {}}` while it should have done `delete a.b`.

For the time being, I'll just comment the `dispatch('deleteMissingProps', _doc)` line in `src/module/actions.ts` and accept that old properties will remain in place (unless I deal with them myself). @mesqueeb We'll see later how to deal with this efficiently if it's ok

**Comment by louisameline** [comment]
It's not well tested **at all**, especially since I don't use fillables and guards, but below is a first attempt to fix the `deleteMissingProps` function. I'll come back if I notice bugs. If you try it, let me know how it works for you. Thanks

```javascript
deleteMissingProps ({getters, commit}, doc) {
  const defaultValues = getters.defaultValues
  const searchTarget = (getters.collectionMode)
    ? getters.storeRef[doc.id]
    : getters.storeRef
  function clearObject(toBeClearedObj, RefObj, path = []) {
    for (let key in toBeClearedObj) {
      const pathCopy = path.slice(0)
      pathCopy.push(key)
      const stringPath = pathCopy.join('.')
      if (!RefObj.hasOwnProperty(key)) {
        const excluded = ['_conf', '_sync']
        if (!excluded.includes(key)
          && (!getters.fillables.length || getters.fillables.includes(stringPath))
          && !getters.guard.includes(stringPath)
        ) {
          commit('DELETE_PROP', (getters.collectionMode ? (doc.id + '.') : '') + stringPath)
        }
      }
      else if (isPlainObject(toBeClearedObj[key])) {
        clearObject(toBeClearedObj[key], RefObj[key], pathCopy);
      }
    }
  }
  clearObject(searchTarget, doc)
},
```

**Comment by mesqueeb** [comment]
@louisameline yeah. This was actually my intended implementation as well. I must have not added sufficient tests, because i'm not sure what's coming through the cracks.
I'll need to set aside some time for this though. Thanks for providing a solution. I'll try to look at your proposal and my own source code asap.
😢
